### PR TITLE
Support deriving stream name from tag

### LINKED ISSUE #67: Derive stream name from fluentd tag
We have an application where we would like to route data to different Kinesis streams based on the fluentd tag. This is doable by specifying a different `<match></match>` configuration for each tag, but it means that there will be a separate KPL process for each configuration. And if we have 20 different streams, that means 20 instances of the KPL process, which wastes resources.

Since the KPL requires the stream name to be specified with each record, the fluentd plugin can dynamically determine the stream name for each record it processes. One possible way to do this would be to allow users to specify a stream name prefix, and a regex to extract other data from the tag. A simpler (and more efficient) alternative would be a simple concatenation.

I'm happy to provide a PR for this functionality, if there is interest in accepting it.

### PR DESCRIPTION
**Pull request by cbroglie**
Add support for kinesis_producer to derive the stream name by concatenating stream_name_prefix (new configuration option) and the fluentd tag. Fixes #67.

### CODE REVIEW

File: lib/fluent/plugin/out_kinesis_producer.rb [hunk 1]
```diff
@@ -20,6 +20,13 @@ class KinesisProducerOutput < BufferedOutput
     Fluent::Plugin.register_output('kinesis_producer', self)
     config_param_for_producer
 
+    def configure(conf)
+      super
+      unless @stream_name or @stream_name_prefix
+        raise Fluent::ConfigError, "'stream_name' or 'stream_name_prefix' is required"
```
Comment (riywo) [inline lib/fluent/plugin/out_kinesis_producer.rb]: How should we handle if user specifies both `stream_name` and `stream_name_prefix`?

**Comment by riywo** [comment]
Hi @cbroglie ,

Thank you for sending a PR. This is so cool and I'd like to merge this one. I've added a small comment to it.

To proceed this, could you try two things?

- Could you confirm this contribution is released under Amazon Software License?
- Could you rebase commits to a single commit?

Best,

File: lib/fluent/plugin/out_kinesis_producer.rb (hunk 1 above)
Comment (cbroglie) [inline lib/fluent/plugin/out_kinesis_producer.rb]: Good point, I'll make that an error as well

**Comment by cbroglie** [comment]
Thanks @riywo. I addressed the comment and rebased the commits. And I can confirm the contribution is released under Amazon Software License.

**Comment by riywo** [comment]
Hi @cbroglie ,

Sorry, I made a mistake on the confirmation; could you confirm that we can use, modify, copy, and redistribute this contribution?

Thanks.

**Comment by cbroglie** [comment]
Yes you may

**Comment by riywo** [comment]
Thank you!

**Pull request merged**